tools/pv_tree_generator/build_tree.py

```diff
@@ -1,243 +1,329 @@
-        ui_node = util.UiNode(pos, {}, False)
-        childStatsVars = []
-        # find all the statsvars belong to the node
-        for sv in stat_vars[pos.key]:
-            if pos.cpv == sv.pv:
-                childStatsVars.append(sv.dcid)
-                root['c'] += 1
-        if len(childStatsVars) > 0:
-            root['cd'].append({
-                'populationType': ui_node.pop_type,
-                'sv': childStatsVars,
-                'l': text_format.format_title(ui_node.text),
-                't': 'v',
-                'c': len(childStatsVars),
-                'mprop': ui_node.mprop,
-            })
-
-    # build specs with >= 1 constraints recursively
-    for pos in pop_obs_spec[1]:
-        child = build_tree_recursive(pos, 1, pop_obs_spec, stat_vars,
-                                     )
-        # For certain branch, we would like to put them under 0 pv nodes:
-        if (pos.pop_type in ['EarthquakeEvent', 'CycloneEvent',
-                             'MortalityEvent']):
-            for pv0 in root['cd']:
-                # hoist logic will break if multiple 0 pv
-                if (pv0['populationType'] == pos.pop_type and pv0['mprop'] == 'count'):
-                    if 'cd' not in pv0:
-                        pv0['cd'] = []
-                    pv0['cd'].append(child)
-                    if 'sv_set' not in pv0:
-                        pv0['sv_set'] = set()
-                    pv0['sv_set'] |= child['sv_set']
-                    break
-        else:
-            root['cd'].append(child)
-        root['sv_set'] |= child['sv_set']
-        del child['sv_set']
+    stats_var_path = {}
+    for v in data:
+        post_process(data[v], [VERTICALS.index(v)], stats_var_path)
+        del data[v]['sv_set']
+    return data, stats_var_path
 
-    # update the count
-    for pv0 in root['cd']:
-        if 'sv_set' in pv0:
-            pv0['c'] += len(pv0['sv_set'])
-            del pv0['sv_set']
-    root['c'] += len(root['sv_set'])
-    del root['sv_set']
-    statsvar_path = {}
-    return traverseTree(root, [vertical_idx], statsvar_path)
+
+def build_tree_recursive(node: PropertyNode, max_level):
+    result = node.json_blob()  # build the property node blobs
+    value_nodes = node.children()  # get the child value nodes
+    for value in value_nodes:
+        # build the child value node blobs
+        value_blob = value.json_blob()
+        result['cd'].append(value_blob)
+        prop_children = value.children()
+        if node.level < max_level:
+            for prop in prop_children:
+                # build the next level
+                prop_blob = build_tree_recursive(prop, max_level)
+                if (prop_blob['cd']):
+                    value_blob['cd'].append(prop_blob)
+    result['cd'] = text_format.filter_and_sort(node.prop, result['cd'], False)
+    return result
 
 
-def traverseTree(root, path, statsvar_path):
+def post_process(root, path, statsvar_path):
```

Comment:
please add function doc string on what processing is in here

Reply:
Done.